src/basic/time-util.c

```diff
@@ -910,14 +908,26 @@ int parse_timestamp(const char *t, usec_t *usec) {
         }
 
         if (pid == 0) {
+                bool with_tz = true;
+
                 if (setenv("TZ", t_timezone, 1) != 0) {
                         shared->return_value = negative_errno();
                         _exit(EXIT_FAILURE);
                 }
 
                 tzset();
 
-                shared->return_value = parse_timestamp_impl(t, &shared->usec, true);
+                /* If there is a timezone that matches the tzname fields, leave the parsing to the implementation
+                 * Otherwise just cut it off */
+                with_tz = !STR_IN_SET(t_timezone, tzname[0], tzname[1]);
+
+                /*cut of the timezone if we dont need it*/
+                if (t_timezone && !with_tz)
+                        t = strdup(t);
+                else
+                        t = strndupa(t, last_space - t);
+
+                shared->return_value = parse_timestamp_impl(t, &shared->usec, with_tz);
```

Comment:
i am not sure what the reasonfor this change is here? why not chop this off unconditionally, after setting TZ to it?

Reply:
The reason about that is that before my patch the tz was always cut off. Which lead to a information loose for example in CET vs. CEST. (The format function gave CET since that was timezone on a given timestamp). Now the parse function takes the current time and uses the rest of the tm function for that, currently i am in CEST, so parse_timestamp_impl would have takes CEST as timezone ... which leads to a lost hour. But if we pass the TZ in t also parse_timestamp_impl and indicate that with the flag, the parsed timestamp does not loose the hour anymore, since the code is handling the timezone correctly.

BUT if the timezone is not specified in the format of CET CEST (which are in tzname[x]) but rather in asian/soeul then the function cannot parse that part, since it is not the part of tzname[x] ... also specified a timezone like asian/soeul looses the information about summer or winter time, so cutting this of to just work with the normal timezone used from now is the most accurate value that could be parsed.

Reply:
Urks, so the issue you are saying that there are two ways to express timezones: the `Europe/Berlin` location specification, and the `CET` and `CEST` timezone specification, and we choke on the latter now?

hmm, this is kinda nasty

Reply:
Yep, well, not too sure if its too nasty, if someone gives a timezone in the notation of "Europe/Berlin" then tzset sets tzname etc.  to the correct timezone, we just dont parse it as we dont get any knowledge over the dayling saving thing. So we can just cut it off here and proceed with the informations gathered form tzset ... :)

Reply:
Sorry for being a bit slow, but I still cannot understand the problem. Can you please provide the exact `*t` and `TZ` that demonstrate the problem.

`CET` is a valid IANA timezone name. So it would go through the fork-based implementation. And it must not touch `tzname` at all.

Reply:
In which condition? A t for a tzname that is different from the timezone? Or a t that triggers the bug of CET vs. CEST in the formated string?

Reply:
Can you please provide all the inputs for `parse_timestamp` and the corresponding `TZ` that demonstrates a problem? I cannot be helpful in this chat until I can reproduce the problem.

Reply:
For the CET vs. CEST:
calling parse_timestamp with
buf = "Sun 1977-05-01 02:03:53 CET"
shows the bug.

The tzname[{0,1}] name vs. the substring in the buf is:
calling parse_timestamp with
buf = "16:20 Asia/Seoul"

Reply:
Now as we have these numbers could you please give me time til tomorrow for a better answer: I'm not having access to my development environment til then (sorry).

Reply:
Meanwhile here is something to think about:

```
$ zdump -v /usr/share/zoneinfo/CET  |grep 1977
/usr/share/zoneinfo/CET  Sun Apr  3 00:59:59 1977 UT = Sun Apr  3 01:59:59 1977 CET isdst=0 gmtoff=3600
/usr/share/zoneinfo/CET  Sun Apr  3 01:00:00 1977 UT = Sun Apr  3 03:00:00 1977 CEST isdst=1 gmtoff=7200
/usr/share/zoneinfo/CET  Sun Sep 25 00:59:59 1977 UT = Sun Sep 25 02:59:59 1977 CEST isdst=1 gmtoff=7200
/usr/share/zoneinfo/CET  Sun Sep 25 01:00:00 1977 UT = Sun Sep 25 02:00:00 1977 CET isdst=0 gmtoff=3600
```
Check the second line.

Reply:
This is valid, timezone changing started in germany in 1980.
Anyway, i think you can take any timezone that is different to the current one to trigger the bug.

Reply:
> This is valid, timezone changing started in germany in 1980.

`CET` is not for Germany, it's a Central European Time, that is not directly connected to Germany and that does not follow the Germany specific timezone rules. For German time one might use `Europe/Berlin` instead.

```
/usr/share/zoneinfo/Europe/Berlin  Sun Oct  2 00:59:59 1949 UT = Sun Oct  2 02:59:59 1949 CEST isdst=1 gmtoff=7200
/usr/share/zoneinfo/Europe/Berlin  Sun Oct  2 01:00:00 1949 UT = Sun Oct  2 02:00:00 1949 CET isdst=0 gmtoff=3600
/usr/share/zoneinfo/Europe/Berlin  Sun Apr  6 00:59:59 1980 UT = Sun Apr  6 01:59:59 1980 CET isdst=0 gmtoff=3600
/usr/share/zoneinfo/Europe/Berlin  Sun Apr  6 01:00:00 1980 UT = Sun Apr  6 03:00:00 1980 CEST isdst=1 gmtoff=7200
```

> Anyway, i think you can take any timezone that is different to the current one to trigger the bug.

I'm in `Pacific/Auckland` and tests for `Asia/Seoul` run for me. So certainly not every timezone.

Reply:
Sorry, i mean: you theoretically can take any date that is in the other daylight saving time, I am in CEST right now, and any date from CET should fail i think, since in the original implementation the timezone from today is taken, since with_tz is always passed as true.

And yes CET is not for Germany, but the country somehow decided that they dont did CET / CEST switching from 1950 to 1979, and formatting dates somehow respects that ... :)

Reply:
I'm not sure it's possible to "fix" it (and personally I don't think it's a bug): the `full timezone name -> abbreviated tz name` is one way conversion and you cannot reuse it to get the same result.

Reply:
@keszybz @poettering sorry for pinging you guys, but we need your help to make a decision.

Reply:
My Proposal: If something like "Asian/Soeul" is specified we just set tm_isdst to something negativ, which means that mktime will also normalize the timezone, and check which one of the possible two is taken, so we have a consistent and correct behaviour IMO.

Reply:
something like = a valid timezone that is different form its tzname[{0;1}]

Reply:
Right, so if you move from Germany to some other nearby country the time specifications would start behave differently.

Ok, I think that might be a compromise.

Reply:
It wouldnt change.
If i pass in `1950-06-06 00:00:00 Asia/Soeul` then tzset with `TZ="Asia/Soeul"` would set tzname[0]=KST (and the other i dont know). 

Then we pass `tm_isdst=-1` which will make `mktime` setting the timezone in the tm struct to the timezone that was valid at `1950-06-06`. And thus we would have a correct timestamp, undependend from the timezone configured on the laptop.